### ISSUE TITLE
Improve on protobuf / alternative serialization formats

### ISSUE THREAD

**Issue by springmeyer** [mapbox/mapnik-vector-tile#1]
Currently we use the protobuf format and the libprotobuf c++ bindings for encoding and decoding vector tiles. But benchmarking has shown with the standard `libprotobuf` API that memory allocation of stl types (mostly std::string) is a bottleneck for parsing messages, even with the help of tcmalloc/jemalloc. One strategy to mitigate this is object re-use but this approach does not mesh well with our target usage from node.js.

Ideas:

 - Recently `capnproto` was released by @kentonv, and the C++ documentation captures perfectly what we have encountered and more: https://github.com/kentonv/capnproto/commit/64286a5fd401a46d245b7f6447871c948f64cf7c
 - We could write our own lazy protobuf parser to avoid incurring the overhead of memory allocation for the temporary protobuf object. The benefit of this could be huge when only some of the layers in the vector tile are needed for rendering (only those used in a given style), or some of the features geometries are requested for rendering based on rules
 - Potentially leverage https://github.com/haberman/upb/, although it appears to have been pulled into internal google development and the [last commit broke](https://github.com/haberman/upb/commit/7d3e2bd2c4cfd1296d1d6f996d7548de26540d41) the public code. 
 - Look into using `set_allocated_foo(Type* foo)`

**Comment by springmeyer** [comment]
Closing this issue - all is well using the protobuf encoding and perf is no longer an issue as we've moved to https://github.com/mapbox/protozero. We could have used `google::protobuf::io::CodedInputStream` but chose to write protozero because:

  - It was difficult to understand how to use of  `google::protobuf::io::CodedInputStream` to achieve what we needed (avoiding allocation of the whole message). With more effort and attention I assume we could have used `google::protobuf::io::CodedInputStream` but made a gut call not to.
  - We desired a header-only library for ease of inclusion in other projects
